Fix for SAPA map presentation

tomaszrybakiewicz opened the pull request:
Closes NAVAND-887

### Description
libanavui-maps:
  - Updated RestAreaProcessor to return SAPA map URL only if the upcoming rest stop is on the current step.

QA-Test-App:
  - Added RestAreaActivity that simulates an example route with SAPA map, and uses MapboxRestAreaApi and MapboxRestAreaGuideMapView
  - Small refactoring to increase code reuse.

### Screenshots or Gifs
_Screen recording of the QA-Test-App  "Rest Area Example" captured on Pixel 6_

https://user-images.githubusercontent.com/2678039/206286049-26ca0b86-9ec7-46d3-a31d-56033838abb2.mp4

### Testing 

> **NOTE:** Loading SAPA maps requires an access token with the correct access level

Before running the QA-Test-App "Rest Area Example", please add a `mapbox_access_token_sapa`  string resource with the SAPA enabled access token to your `qa-test-app/src/main/res/values/mapbox_access_token.xml` file:

```xml
<string name="mapbox_access_token_sapa" translatable="false">SAPA_ACCESS_TOKEN</string>
```
Please contact me via SLACK for the access token.

abhishek1508 commented on qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/componentinstaller/components/RestAreaGuideMap.kt:
```diff
@@ -0,0 +1,36 @@
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        mapboxNavigation.flowRouteProgress().observe { routeProgress ->
+            restAreaApi.generateUpcomingRestAreaGuideMap(routeProgress) {
+                restAreaView.render(it)
+            }
+        }
```
`RestStop` can be embedded either in `route.leg.step.bannerInstructions` or `route.leg.step.intersections()`. Hence we have 2 APIs supported in `MapboxRestAreaApi` as follows 

using bannerInstructions
```
fun generateRestAreaGuideMap(
    instructions: BannerInstructions,
    consumer: MapboxNavigationConsumer<Expected<RestAreaGuideMapError, 
    RestAreaGuideMapValue>>
)
```

using RouteProgress
```
fun generateUpcomingRestAreaGuideMap(
    routeProgress: RouteProgress,
    consumer: MapboxNavigationConsumer<Expected<RestAreaGuideMapError, 
    RestAreaGuideMapValue>>
)
```

This is a nice component to have, but based on the logic it only renders `rest_stop.guideMap` if the data is available via `generateUpcomingRestAreaGuideMap(routeProgress)`. We should modify the logic to make it so that it can render rest stops when it is embedded in banner instructions as well. 

The logic can be changed as following:

```
mapboxNavigation.flowRouteProgress().observe { routeProgress ->
    val progressJob = async {
        routeProgress.let { restAreaApi.generateUpcomingRestAreaGuideMap(it) }
    }
    val progressJob = async {
        routeProgress.bannerInstructions?.let { restAreaApi.generateRestAreaGuideMap(it) }
    }
    val progressResult = progressJob.await()
    val result = if (progressResult.isValue) {
        instructionsJob.cancel()
        progressResult
    } else {
        instructionsJob.await()
    }
    result?.let {
        restAreaView.render(it)
    }
}

private suspend fun MapboxRestAreaApi.generateUpcomingRestAreaGuideMap(
    progress: RouteProgress,
): Expected<RestAreaGuideMapError, RestAreaGuideMapValue> {
    return suspendCancellableCoroutine { generateUpcomingRestAreaGuideMap(progress, it::resume) }
}

private suspend fun MapboxRestAreaApi.generateRestAreaGuideMap(
    instructions: BannerInstructions,
): Expected<RestAreaGuideMapError, RestAreaGuideMapValue> {
    return suspendCancellableCoroutine { generateRestAreaGuideMap(instructions, it::resume) }
}
```
Once we update the logic as above, we can use the same activity to test routes that
- contains rest stop in bannerInstructions only
- contains rest stop in step.intersections only
- contains rest stop in both bannerInstructions and step.intersections

We could probably change the spinner to select routes containing rest stops based on bannerInstructions or step.intersections or both instead of short/long SAPA route.

abhishek1508 replied:
Won't do. To be picked as follow-up.

abhishek1508 commented on qa-test-app/src/main/res/layout/layout_activity_rest_area.xml:
```diff
@@ -0,0 +1,80 @@
+      app:layout_constraintTop_toTopOf="parent"
+      />
+
+  <FrameLayout
+      android:layout_width="150dp"
+      android:layout_height="100dp"
+      android:layout_marginStart="4dp"
+      android:layout_marginTop="50dp"
+      android:background="#33000000"
+      app:layout_constraintStart_toStartOf="parent"
+      app:layout_constraintTop_toTopOf="parent"
+      >
+
+    <androidx.appcompat.widget.AppCompatTextView
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:textAlignment="center"
+        android:text="SAPA map\nnot available"
+        />
+
+    <com.mapbox.navigation.ui.maps.guidance.restarea.view.MapboxRestAreaGuideMapView
+        android:id="@+id/restAreaView"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        />
+  </FrameLayout>
```
I don't think we need this textview to show `SAPA map not available`. I would vote to show the SAPA map when one is available at the current step otherwise don't show any UI and navigate as regular. The text can be treated in 2 different forms, "is it not available for the complete route" or "is it not available at the current step" and this can lead to some confusion

tomaszrybakiewicz commented on qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/view/componentinstaller/components/MapMarkersRestStops.kt:
```diff
@@ -0,0 +1,59 @@
+        super.onAttached(mapboxNavigation)
+
+        val icon = IconFactory(mapView.context).pinIconWithText(
+            "zzZ",
```
That's a matter of preference. I've created this debug component to help me visualize all upcoming REST STOPs. The "zzZ" marker serves that purpose and is good enough.

tomaszrybakiewicz commented:
@abhishek1508 Updated. I've replaced `MapboxNavigationProvider.create(...)` usage with the `requireMapboxNavigation(...)` delegate, updated the "text marker" drawable to accommodate larger text and updated RestStop markers to say "REST STOP".
Please let me know if there is anything else you would like to change.

tomaszrybakiewicz commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/RestAreaProcessor.kt:
```diff
@@ -116,18 +111,31 @@ internal object RestAreaProcessor {
-    ): UpcomingRoadObject? {
-        return this.upcomingRoadObjects.firstOrNull {
-            if (it.roadObject.objectType == type) {
+    private fun RouteProgress.getUpcomingRestStopOnCurrentStep(): RestStop? {
+        val upcomingRestStop = getFirstUpcomingRestStop() ?: return null
+        val upcomingRestStopLocation = upcomingRestStop.location.shape as? Point ?: return null
+
+        val currentStep = currentLegProgress?.currentStepProgress?.step
+        val isUpcomingRestStopAtCurrentStep =
+            !currentStep?.getRestStopsMatchingLocation(upcomingRestStopLocation).isNullOrEmpty()
+
+        return if (isUpcomingRestStopAtCurrentStep) upcomingRestStop else null
+    }
```
> What happens when we have overlapping geometries?

What do you mean? Like two `rest_stops` at the same intersection location?

abhishek1508 replied:
No I mean use case when the route goes through the same rest stop more than 1 time.

tomaszrybakiewicz replied:
I see. The `getFirstUpcomingRestStop()` method iterates through `RouteProgress.upcomingRoadObjects` and returns the first `RestStop` road object, which has `distanceToStart > 0`. The `RouteProgress.upcomingRoadObjects` list originates from the `NavigationStatus.getUpcomingRouteAlerts()`. The list is created by the Native Navigator and passed to the NavSDK (`MapboxTripSession`) via `NavigatorObserver` as a member of `NavigationStatus`. 
So... we will show the same rest stop map as many times as it's in the `NavigationStatus.getUpcomingRouteAlerts()` list.

LukasPaczos replied:
If we have a route which passes through the same section of the road twice, the data about that road section will be in the response twice as well, and so will the road objects. If that helps @abhishek1508 with your concern.

abhishek1508 replied:
I understand that, but I was more wondering if this logic could fail and maybe we should distance calculations as discussed earlier. But, I see you already added this [comment](https://github.com/mapbox/mapbox-navigation-android/pull/6695#discussion_r1044527857).

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/guidance/restarea/RestAreaProcessor.kt:
```diff
@@ -116,18 +111,31 @@ internal object RestAreaProcessor {
+        val upcomingRestStop = getFirstUpcomingRestStop() ?: return null
+        val upcomingRestStopLocation = upcomingRestStop.location.shape as? Point ?: return null
+
+        val currentStep = currentLegProgress?.currentStepProgress?.step
+        val isUpcomingRestStopAtCurrentStep =
+            !currentStep?.getRestStopsMatchingLocation(upcomingRestStopLocation).isNullOrEmpty()
```
```suggestion
        val currentStepProgress = currentLegProgress?.currentStepProgress ?: return null
        val isUpcomingRestStopAtCurrentStep = upcomingRestStop.distanceToStart - currentStepProgress.distanceRemaining <= 0
```
This still looks to me like an approach worth considering to avoid point precision issues and just speed up the logic.

tomaszrybakiewicz replied:
Great idea. Thanks.

tomaszrybakiewicz replied:
Done. Please update version.

abhishek1508 reviewed (approved):
Thanks for addressing the feedback.
